### DISCUSSION TITLE
Call custom function into store method

### DISCUSSION THREAD

**Discussion by ycgarrido** [pmndrs/zustand#1836] [Q&A] [answered]
## Summary

I have a `useAlert` hook that returns a `show` function.

## Link to reproduction

```jsx
// Component.tsx
const Component = ()=>{
  const { show } = useAlert();
  const { updateAction } = useStore();

  ....
}

// store.ts
const useStore = create<MyType>((set, get) => ({
  updateAction: async()=>{
     const result = await fetch(...);
     show("My success message"); 
 }
}));
```
How can I call the show function inside my store? I have tried in a thousand ways but I have not been able to.

**Comment by dbritto-dev** [comment, marked as answer]
@ycgarrido there're many ways to achieve what you want. Here're couple examples:

```ts
// assuming `useAlert` is a zustand store too

const useStore = create<MyType>()(() => ({
  updateAction: async () => {
    const result = await fetch(...)
    useAlert.getState().show('My success message')
  }
}))  
```

```ts
const useStore = create<MyType>()(() => ({
  updateAction: async ({ alert }) => {
    const result = await fetch(...)
    alert.show('My success message')
  }
}))  

const alert = useAlert()
const store = useStore()

store.updateAction({ alert })
```

**Reply by dai-shi** [reply]
The first one doesn't seem very good as it can create a new store on every render.

**Reply by dbritto-dev** [reply]
@dai-shi you are right, what would you do in that case?

**Reply by dbritto-dev** [reply]
@dai-shi I just left the other ones

**Reply by samhirtarif** [reply]
Would any of these even be doable or am I missing something? It is stated that `useAlert` is a hook and it shouldn't be callable from anywhere other than the top level of a component 🤔

**Reply by samhirtarif** [reply]
One way I can think of doing it if it is a hook is passing the `show` method as a callback to the `updateAction`

```js
// Component.tsx
const Component = ()=>{
  const { show } = useAlert();
  const { updateAction } = useStore();

  ....
  updateAction(show)
}

// store.ts
const useStore = create<MyType>((set, get) => ({
  updateAction: async (callback)=>{
     const result = await fetch(...);
     callback?.("My success message"); 
 }
}));
```
Something along these lines

**Reply by dbritto-dev** [reply]
@samhirtarif yeah, they're doable I just removed the one that use `useAlert` and create `useStore` in a custom hooks since lead us to create new stores on every render I just left the ones that you can use

**Comment by dbritto-dev** [comment]
@ycgarrido let me know if that works for you

**Reply by ycgarrido** [reply]
Yes, works for me, Thanks!!!
